nav2_dwb_controller/dwb_core/src/publisher.cpp

```diff
@@ -256,46 +257,84 @@ DWBPublisher::publishCostGrid(
 
   if (!publish_cost_grid_pc_) {return;}
 
-  auto cost_grid_pc = std::make_unique<sensor_msgs::msg::PointCloud>();
+  auto cost_grid_pc = std::make_unique<sensor_msgs::msg::PointCloud2>();
   cost_grid_pc->header.frame_id = costmap_ros->getGlobalFrameID();
   cost_grid_pc->header.stamp = clock_->now();
 
   nav2_costmap_2d::Costmap2D * costmap = costmap_ros->getCostmap();
   double x_coord, y_coord;
   unsigned int size_x = costmap->getSizeInCellsX();
   unsigned int size_y = costmap->getSizeInCellsY();
-  cost_grid_pc->points.resize(size_x * size_y);
-  unsigned int i = 0;
-  for (unsigned int cy = 0; cy < size_y; cy++) {
-    for (unsigned int cx = 0; cx < size_x; cx++) {
-      costmap->mapToWorld(cx, cy, x_coord, y_coord);
-      cost_grid_pc->points[i].x = x_coord;
-      cost_grid_pc->points[i].y = y_coord;
-      i++;
-    }
-  }
 
-  sensor_msgs::msg::ChannelFloat32 totals;
-  totals.name = "total_cost";
-  totals.values.resize(size_x * size_y, 0.0);
+std::vector<std::pair<std::string, std::vector<float>>> cost_channels;
+  std::vector<float> total_cost(size_x * size_y, 0.0);
 
   for (TrajectoryCritic::Ptr critic : critics) {
-    unsigned int channel_index = cost_grid_pc->channels.size();
-    critic->addCriticVisualization(*cost_grid_pc);
-    if (channel_index == cost_grid_pc->channels.size()) {
+    unsigned int channel_index = cost_channels.size();
+    critic->addCriticVisualization(cost_channels);
+    if (channel_index == cost_channels.size()) {
       // No channels were added, so skip to next critic
       continue;
     }
     double scale = critic->getScale();
-    for (i = 0; i < size_x * size_y; i++) {
-      totals.values[i] += cost_grid_pc->channels[channel_index].values[i] * scale;
+    for (unsigned int i = 0; i < size_x * size_y; i++) {
+      total_cost[i] += cost_channels[channel_index].second[i] * scale;
+    }
+  }
+
+  cost_channels.push_back(std::make_pair("total_cost", total_cost));
+
+  cost_grid_pc->width = size_x * size_y;  // Should this be height and width instead ?
```

Comment:
You have some of these comments that seem like questions - we should resolve these + remove them before merging